File: backend/models/roles.py
"""
User roles and permissions enums
"""
from enum import Enum
from typing import List, Set


class UserRole(str, Enum):
    """
    Hiérarchie des rôles utilisateurs
    """
    ADMIN = "ADMIN"  # Admin/Account Owner - Tous les droits
    CHEF_OPERATOR = "CHEF_OPERATOR"  # Chef Operator - Gestion équipe + training
    OPERATOR = "OPERATOR"  # Operator - Annotation + inférence basique
    VIEWER = "VIEWER"  # Viewer - Lecture seule (legacy)


class Permission(str, Enum):
    """
    Permissions granulaires du système
    """
    # Gestion des utilisateurs
    USER_READ = "user:read"
    USER_CREATE = "user:create"
    USER_UPDATE = "user:update"
    USER_DELETE = "user:delete"
    USER_PROMOTE = "user:promote"  # Changer les rôles
    
    # Gestion des projets
    PROJECT_READ = "project:read"
    PROJECT_CREATE = "project:create"
    PROJECT_UPDATE = "project:update"
    PROJECT_DELETE = "project:delete"
    
    # Gestion des images
    IMAGE_READ = "image:read"
    IMAGE_UPLOAD = "image:upload"
    IMAGE_DELETE = "image:delete"
    
    # Annotations
    ANNOTATION_READ = "annotation:read"
    ANNOTATION_CREATE = "annotation:create"
    ANNOTATION_UPDATE = "annotation:update"
    ANNOTATION_DELETE = "annotation:delete"
    ANNOTATION_VALIDATE = "annotation:validate"  # Valider les annotations des autres
    
    # Training de modèles
    TRAINING_READ = "training:read"
    TRAINING_START = "training:start"
    TRAINING_CANCEL = "training:cancel"
    TRAINING_DELETE = "training:delete"
    
    # Modèles
    MODEL_READ = "model:read"
    MODEL_DOWNLOAD = "model:download"
    MODEL_DEPLOY = "model:deploy"  # Mettre en production
    MODEL_DELETE = "model:delete"
    
    # Inférence
    INFERENCE_RUN = "inference:run"
    INFERENCE_READ = "inference:read"
    INFERENCE_DELETE = "inference:delete"
    
    # Analytics & Reports
    ANALYTICS_READ = "analytics:read"
    REPORTS_GENERATE = "reports:generate"
    
    # Settings & Configuration
    SETTINGS_READ = "settings:read"
    SETTINGS_UPDATE = "settings:update"


# Matrice des permissions par rôle
ROLE_PERMISSIONS: dict[UserRole, Set[Permission]] = {
    UserRole.ADMIN: {
        # Tous les droits (Admin/Account Owner)
        *Permission.__members__.values()
    },
    
    UserRole.CHEF_OPERATOR: {
        # Gestion d'équipe limitée
        Permission.USER_READ,
        Permission.USER_UPDATE,  # Peut modifier info des operators
        
        # Gestion complète des projets
        Permission.PROJECT_READ,
        Permission.PROJECT_CREATE,
        Permission.PROJECT_UPDATE,
        Permission.PROJECT_DELETE,
        
        # Gestion complète des images
        Permission.IMAGE_READ,
        Permission.IMAGE_UPLOAD,
        Permission.IMAGE_DELETE,
        
        # Gestion complète des annotations
        Permission.ANNOTATION_READ,
        Permission.ANNOTATION_CREATE,
        Permission.ANNOTATION_UPDATE,
        Permission.ANNOTATION_DELETE,
        Permission.ANNOTATION_VALIDATE,
        
        # Gestion complète du training
        Permission.TRAINING_READ,
        Permission.TRAINING_START,
        Permission.TRAINING_CANCEL,
        Permission.TRAINING_DELETE,
        
        # Gestion des modèles
        Permission.MODEL_READ,
        Permission.MODEL_DOWNLOAD,
        Permission.MODEL_DEPLOY,
        Permission.MODEL_DELETE,
        
        # Inférence complète
        Permission.INFERENCE_RUN,
        Permission.INFERENCE_READ,
        Permission.INFERENCE_DELETE,
        
        # Analytics
        Permission.ANALYTICS_READ,
        Permission.REPORTS_GENERATE,
        
        # Settings lecture seule
        Permission.SETTINGS_READ,
    },
    
    UserRole.OPERATOR: {
        # Lecture des utilisateurs
        Permission.USER_READ,
        
        # Projets (lecture + création limitée)
        Permission.PROJECT_READ,
        
        # Images (upload et lecture)
        Permission.IMAGE_READ,
        Permission.IMAGE_UPLOAD,
        
        # Annotations (CRUD de ses propres annotations)
        Permission.ANNOTATION_READ,
        Permission.ANNOTATION_CREATE,
        Permission.ANNOTATION_UPDATE,
        Permission.ANNOTATION_DELETE,
        
        # Training (lecture seule)
        Permission.TRAINING_READ,
        
        # Modèles (lecture + téléchargement)
        Permission.MODEL_READ,
        Permission.MODEL_DOWNLOAD,
        
        # Inférence (run + lecture)
        Permission.INFERENCE_RUN,
        Permission.INFERENCE_READ,
        
        # Analytics (lecture seule)
        Permission.ANALYTICS_READ,
    },
    
    UserRole.VIEWER: {
        # Lecture seule (legacy)
        Permission.USER_READ,
        Permission.PROJECT_READ,
        Permission.IMAGE_READ,
        Permission.ANNOTATION_READ,
        Permission.TRAINING_READ,
        Permission.MODEL_READ,
        Permission.INFERENCE_READ,
        Permission.ANALYTICS_READ,
    }
}


def get_permissions_for_role(role: UserRole) -> Set[Permission]:
    """Récupérer toutes les permissions pour un rôle donné"""
    return ROLE_PERMISSIONS.get(role, set())


def has_permission(role: UserRole, permission: Permission) -> bool:
    """Vérifier si un rôle possède une permission spécifique"""
    return permission in get_permissions_for_role(role)


def can_promote_to_role(promoter_role: UserRole, target_role: UserRole) -> bool:
    """
    Vérifier si un utilisateur peut promouvoir vers un rôle
    Règles :
    - ADMIN peut promouvoir vers tous les rôles
    - CHEF_OPERATOR ne peut pas promouvoir (seulement Admin)
    - OPERATOR ne peut pas promouvoir
    """
    if promoter_role == UserRole.ADMIN:
        return True
    return False


def get_role_hierarchy() -> List[UserRole]:
    """Retourner la hiérarchie des rôles (du plus haut au plus bas)"""
    return [
        UserRole.ADMIN,
        UserRole.CHEF_OPERATOR,
        UserRole.OPERATOR,
        UserRole.VIEWER
    ]


def get_role_display_name(role: UserRole) -> str:
    """Nom d'affichage du rôle en français"""
    display_names = {
        UserRole.ADMIN: "Administrateur",
        UserRole.CHEF_OPERATOR: "Chef Opérateur",
        UserRole.OPERATOR: "Opérateur",
        UserRole.VIEWER: "Observateur"
    }
    return display_names.get(role, role.value)


def get_role_description(role: UserRole) -> str:
    """Description du rôle"""
    descriptions = {
        UserRole.ADMIN: "Tous les droits sur la plateforme. Gestion des utilisateurs, projets, et configuration système.",
        UserRole.CHEF_OPERATOR: "Gestion des projets, équipe, training et déploiement des modèles. Supervision des annotations.",
        UserRole.OPERATOR: "Annotation d'images, exécution d'inférences. Accès limité aux projets assignés.",
        UserRole.VIEWER: "Accès en lecture seule à l'ensemble de la plateforme."
    }
    return descriptions.get(role, "")

File: frontend/src/types/roles.ts
/**
 * Types et définitions pour le système de rôles et permissions
 * Correspond aux définitions du backend (backend/models/roles.py)
 */

export enum UserRole {
  ADMIN = "ADMIN",
  CHEF_OPERATOR = "CHEF_OPERATOR",
  OPERATOR = "OPERATOR",
  VIEWER = "VIEWER",
}

export enum Permission {
  // Projets
  PROJECT_CREATE = "PROJECT_CREATE",
  PROJECT_READ = "PROJECT_READ",
  PROJECT_UPDATE = "PROJECT_UPDATE",
  PROJECT_DELETE = "PROJECT_DELETE",
  PROJECT_ARCHIVE = "PROJECT_ARCHIVE",

  // Images
  IMAGE_UPLOAD = "IMAGE_UPLOAD",
  IMAGE_READ = "IMAGE_READ",
  IMAGE_UPDATE = "IMAGE_UPDATE",
  IMAGE_DELETE = "IMAGE_DELETE",
  IMAGE_BULK_DELETE = "IMAGE_BULK_DELETE",

  // Annotations
  ANNOTATION_CREATE = "ANNOTATION_CREATE",
  ANNOTATION_READ = "ANNOTATION_READ",
  ANNOTATION_UPDATE = "ANNOTATION_UPDATE",
  ANNOTATION_DELETE = "ANNOTATION_DELETE",
  ANNOTATION_REVIEW = "ANNOTATION_REVIEW",
  ANNOTATION_VALIDATE = "ANNOTATION_VALIDATE",

  // Modèles
  MODEL_READ = "MODEL_READ",
  MODEL_UPLOAD = "MODEL_UPLOAD",
  MODEL_DELETE = "MODEL_DELETE",
  MODEL_DEPLOY = "MODEL_DEPLOY",

  // Entraînement
  TRAINING_START = "TRAINING_START",
  TRAINING_READ = "TRAINING_READ",
  TRAINING_CANCEL = "TRAINING_CANCEL",
  TRAINING_DELETE = "TRAINING_DELETE",

  // Inférence
  INFERENCE_RUN = "INFERENCE_RUN",
  INFERENCE_READ = "INFERENCE_READ",
  INFERENCE_DELETE = "INFERENCE_DELETE",

  // XAI (Explainability)
  XAI_GENERATE = "XAI_GENERATE",
  XAI_READ = "XAI_READ",

  // Feedback
  FEEDBACK_CREATE = "FEEDBACK_CREATE",
  FEEDBACK_READ = "FEEDBACK_READ",
  FEEDBACK_RESOLVE = "FEEDBACK_RESOLVE",

  // Rapports
  REPORT_GENERATE = "REPORT_GENERATE",
  REPORT_READ = "REPORT_READ",
  REPORT_EXPORT = "REPORT_EXPORT",

  // Utilisateurs (Admin uniquement)
  USER_READ = "USER_READ",
  USER_CREATE = "USER_CREATE",
  USER_UPDATE = "USER_UPDATE",
  USER_DELETE = "USER_DELETE",
  USER_MANAGE_ROLES = "USER_MANAGE_ROLES",

  // Audit
  AUDIT_READ = "AUDIT_READ",
}

// Matrice des permissions par rôle (correspond au backend)
export const ROLE_PERMISSIONS: Record<UserRole, Permission[]> = {
  [UserRole.ADMIN]: Object.values(Permission), // Toutes les permissions

  [UserRole.CHEF_OPERATOR]: [
    // Projets
    Permission.PROJECT_CREATE,
    Permission.PROJECT_READ,
    Permission.PROJECT_UPDATE,
    Permission.PROJECT_DELETE,
    Permission.PROJECT_ARCHIVE,
    // Images
    Permission.IMAGE_UPLOAD,
    Permission.IMAGE_READ,
    Permission.IMAGE_UPDATE,
    Permission.IMAGE_DELETE,
    Permission.IMAGE_BULK_DELETE,
    // Annotations
    Permission.ANNOTATION_CREATE,
    Permission.ANNOTATION_READ,
    Permission.ANNOTATION_UPDATE,
    Permission.ANNOTATION_DELETE,
    Permission.ANNOTATION_REVIEW,
    Permission.ANNOTATION_VALIDATE,
    // Modèles
    Permission.MODEL_READ,
    Permission.MODEL_UPLOAD,
    Permission.MODEL_DELETE,
    Permission.MODEL_DEPLOY,
    // Entraînement
    Permission.TRAINING_START,
    Permission.TRAINING_READ,
    Permission.TRAINING_CANCEL,
    Permission.TRAINING_DELETE,
    // Inférence
    Permission.INFERENCE_RUN,
    Permission.INFERENCE_READ,
    Permission.INFERENCE_DELETE,
    // XAI
    Permission.XAI_GENERATE,
    Permission.XAI_READ,
    // Feedback
    Permission.FEEDBACK_CREATE,
    Permission.FEEDBACK_READ,
    Permission.FEEDBACK_RESOLVE,
    // Rapports
    Permission.REPORT_GENERATE,
    Permission.REPORT_READ,
    Permission.REPORT_EXPORT,
    // Audit
    Permission.AUDIT_READ,
  ],

  [UserRole.OPERATOR]: [
    // Projets (lecture seule)
    Permission.PROJECT_READ,
    // Images
    Permission.IMAGE_UPLOAD,
    Permission.IMAGE_READ,
    Permission.IMAGE_UPDATE,
    Permission.IMAGE_DELETE,
    // Annotations
    Permission.ANNOTATION_CREATE,
    Permission.ANNOTATION_READ,
    Permission.ANNOTATION_UPDATE,
    Permission.ANNOTATION_DELETE,
    // Modèles (lecture seule)
    Permission.MODEL_READ,
    // Entraînement
    Permission.TRAINING_START,
    Permission.TRAINING_READ,
    // Inférence
    Permission.INFERENCE_RUN,
    Permission.INFERENCE_READ,
    // XAI
    Permission.XAI_GENERATE,
    Permission.XAI_READ,
    // Feedback
    Permission.FEEDBACK_CREATE,
    Permission.FEEDBACK_READ,
    // Rapports (lecture seule)
    Permission.REPORT_READ,
  ],

  [UserRole.VIEWER]: [
    // Projets (lecture seule)
    Permission.PROJECT_READ,
    // Images (lecture seule)
    Permission.IMAGE_READ,
    // Annotations (lecture seule)
    Permission.ANNOTATION_READ,
    // Modèles (lecture seule)
    Permission.MODEL_READ,
    // Entraînement (lecture seule)
    Permission.TRAINING_READ,
    // Inférence (lecture seule)
    Permission.INFERENCE_READ,
    // XAI (lecture seule)
    Permission.XAI_READ,
    // Rapports (lecture seule)
    Permission.REPORT_READ,
  ],
};

// Hiérarchie des rôles (pour les promotions)
export const ROLE_HIERARCHY: Record<UserRole, number> = {
  [UserRole.VIEWER]: 0,
  [UserRole.OPERATOR]: 1,
  [UserRole.CHEF_OPERATOR]: 2,
  [UserRole.ADMIN]: 3,
};

// Badges de couleur par rôle
export const ROLE_COLORS: Record<UserRole, string> = {
  [UserRole.ADMIN]: "bg-violet-100 text-violet-600 border-violet-200",
  [UserRole.CHEF_OPERATOR]: "bg-indigo-100 text-indigo-600 border-indigo-200",
  [UserRole.OPERATOR]: "bg-emerald-100 text-emerald-600 border-emerald-200",
  [UserRole.VIEWER]: "bg-slate-100 text-slate-600 border-slate-200",
};

// Labels lisibles pour les rôles
export const ROLE_LABELS: Record<UserRole, string> = {
  [UserRole.ADMIN]: "Administrateur",
  [UserRole.CHEF_OPERATOR]: "Chef Opérateur",
  [UserRole.OPERATOR]: "Opérateur",
  [UserRole.VIEWER]: "Visualiseur",
};

// Descriptions des rôles
export const ROLE_DESCRIPTIONS: Record<UserRole, string> = {
  [UserRole.ADMIN]: "Accès complet à toutes les fonctionnalités, gestion des utilisateurs",
  [UserRole.CHEF_OPERATOR]: "Gestion complète des projets, modèles et entraînements",
  [UserRole.OPERATOR]: "Annotation, entraînement et inférence sur les projets",
  [UserRole.VIEWER]: "Consultation des projets, modèles et résultats",
};

/**
 * Vérifier si un utilisateur a une permission
 */
export function hasPermission(userRole: UserRole, permission: Permission): boolean {
  return ROLE_PERMISSIONS[userRole].includes(permission);
}

/**
 * Vérifier si un utilisateur a toutes les permissions spécifiées
 */
export function hasAllPermissions(userRole: UserRole, permissions: Permission[]): boolean {
  return permissions.every(p => hasPermission(userRole, p));
}

/**
 * Vérifier si un utilisateur a au moins une des permissions spécifiées
 */
export function hasAnyPermission(userRole: UserRole, permissions: Permission[]): boolean {
  return permissions.some(p => hasPermission(userRole, p));
}

/**
 * Vérifier si un rôle peut être promu vers un autre rôle
 */
export function canPromoteTo(currentRole: UserRole, targetRole: UserRole): boolean {
  return ROLE_HIERARCHY[targetRole] > ROLE_HIERARCHY[currentRole];
}

/**
 * Vérifier si un rôle peut être rétrogradé vers un autre rôle
 */
export function canDemoteTo(currentRole: UserRole, targetRole: UserRole): boolean {
  return ROLE_HIERARCHY[targetRole] < ROLE_HIERARCHY[currentRole];
}

/**
 * Obtenir toutes les permissions d'un rôle
 */
export function getPermissionsForRole(role: UserRole): Permission[] {
  return ROLE_PERMISSIONS[role];
}

/**
 * Obtenir les rôles disponibles pour la promotion (supérieurs au rôle actuel)
 */
export function getPromotableRoles(currentRole: UserRole): UserRole[] {
  const currentLevel = ROLE_HIERARCHY[currentRole];
  return Object.entries(ROLE_HIERARCHY)
    .filter(([_, level]) => level > currentLevel)
    .map(([role, _]) => role as UserRole);
}

/**
 * Obtenir les rôles disponibles pour la rétrogradation (inférieurs au rôle actuel)
 */
export function getDemotableRoles(currentRole: UserRole): UserRole[] {
  const currentLevel = ROLE_HIERARCHY[currentRole];
  return Object.entries(ROLE_HIERARCHY)
    .filter(([_, level]) => level < currentLevel)
    .map(([role, _]) => role as UserRole);
}
